docs/project/cpp_style_guide.md

```diff
@@ -151,6 +152,12 @@ these.
         `namespace Carbon::Testing { namespace { ... } }` to keep everything
         internal.
 
+### Declaration order
+
+-   For a struct without constructors or factory functions, fields should be
+    declared at the start of the struct, immediately after any member types or
+    typedefs, not at the end.
```

Comment:
If we want to match the upstream direction, we should use the style guide's existing distinction between a "struct" and a "class" which provides most of the nuance needed here:

```suggestion
-   For a [struct as opposed to a class](https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes), fields should be
    declared at the start of the struct, immediately after any public member types or
    typedefs.
```

Notably, adding constructors or factory functions are all fine. It is only introducing a *type invariant* that would trigger a change, and it would also trigger a change to use the `class` keyword as well as other changes.

Reply:
My concern here is that if a struct starts as:

```
struct Foo {
  int a;
  std::string b;

  // ... other stuff ...
```

that people will try to initialize it as `Foo{.a = 1, .b = "hello"}` and be surprised if that fails. And it will fail if `Foo` also has a constructor. Similarly if `Foo` has factory functions, people might not use them when appropriate if aggregate initialization is available and the members precede the factory functions. I'd be fine with this ordering:

- member types
- constructors and factory functions
- public data members
- public member functions
- ...

but that's not consistent with the new rule in the baseline style guide.